lib/ibandit/german_details_converter.rb

```diff
@@ -0,0 +1,1270 @@
+# German bank details don't map directly to IBANs in the same way as in other
+# countries - each bank has idiosyncracies for translating its cusomers' bank
+# details. These idiosyncracies are described in a document from the Bundesbank:
+# http://www.bundesbank.de/Navigation/DE/Aufgaben/Unbarer_Zahlungsverkehr/IBAN_Regeln/iban_regeln.html
+
+module Ibandit
+  module GermanDetailsConverter
+    def self.rules
+      @rules ||= YAML.load_file(
+        File.expand_path('../../../data/german_iban_rules.yml', __FILE__)
+      )
+    end
+
+    def self.convert(opts)
+      # Fetch the relevant rule number. Default to '000000'
+      rule_num = rules.fetch(opts[:bank_code], {}).fetch(:iban_rule, '000000')
+
+      # Convert the bank details using the relevant rule
+      updated_bank_details = const_get("Rule#{rule_num}").new(
+        opts[:bank_code],
+        opts[:account_number]
+      ).converted_details
+
+      opts.merge(updated_bank_details)
+    end
+
+    ##############
+    # IBAN Rules #
+    ##############
+
+    class BaseRule
+      def initialize(bank_code, account_number)
+        @bank_code = bank_code
+        @account_number = account_number
+      end
+
+      def converted_details
+        raise NotImplementedError, 'Concrete RuleXXXXXX classes should ' \
+                                   'define a converted_details function'
+      end
+    end
+
+    module PseudoAccountNumberBehaviour
+      def converted_details
+        updated_account_number =
+          self.class.const_get('PSEUDO_ACCOUNT_NUMBER_MAPPING').fetch(
```

Comment:
As I mentioned could use a class instance variable approach here to avoid the use of reflection. That would also allow for subclasses overriding the definition of the variable if required.